pydeseq2/utils.py

```diff
@@ -1608,3 +1666,226 @@ def lowess(
+
+    Parameters
+    ----------
+    colname : str
+        Column name to parse.
+    split_interactions : bool
+        Whether to split interactions terms (of the form "var1:var2").
+        (default: ``True``).
+
+    Returns
+    -------
+    str or list
+        Parsed factor(s).
+    """
+    # There can be either 0 or one vs
+    factors = colname.split("_vs_")
+    assert len(factors) in {1, 2}, f"Invalid column name: {colname}."
+    if len(factors) == 1:
+        return [factors[0]]
+    else:
+        # Left part before the underscore is an interaction term
+        name_with_interaction = factors[0].split("_")[0]
+        if not split_interactions:
+            return name_with_interaction
+        else:
+            return name_with_interaction.split(":")
+
+
+def process_design_factors(
+    design_factors: str | list,
+    metadata_columns: pd.Index,
+    continuous_factors: list[str] | None = None,
+    ref_level: list[tuple[str, str]] | None = None,
+    design_matrix: pd.DataFrame | None = None,
+):
+    """Process design factors.
+
+    Convert design factors to a formulaic string and check validity.
+
+    Parameters
+    ----------
+    design_factors : str or list
+        Design factors to process. If a list, it will be converted to a formulaic
+        string. If a string, it will be checked for validity.
+
+    metadata_columns : pandas.Index
+        Metadata columns to check against (obtained from .obs).
+
+    continuous_factors : list of str, optional
+        List of continuous factors. If provided, underscores will be replaced by
+        hyphens.
+
+    ref_level : str, optional
+        Reference level. If provided, underscores will be replaced by hyphens.
+
+    design_matrix : pd.DataFrame, optional
+        Design matrix. If provided, design factors will be extracted from it.
+
+    Returns
+    -------
+    design_factors : str
+        Formulaic string of design factors.
+
+    design_factors_list : list
+        List of design factors. May contain interaction terms.
+
+    single_design_factors : list
+        List of individual design factors.
+
+    continuous_factors : list of str or None
+        List of continuous factors with underscores replaced by hyphens.
+
+    ref_level : str or None
+        Reference level with underscores replaced by hyphens.
+
+    """
+    # If design factors is provided as a list, convert it to a formulaic string
+    if isinstance(design_factors, list):
+        # Remove space around colons
+        design_factors = [
+            re.sub(r"(?:(?<=\:) +| +(?=\:))", "", factor) for factor in design_factors
+        ]
+        # Extract single factors
+        extracted_single_factors = list(
+            itertools.chain(*[factor.split(":") for factor in design_factors])
+        )
+
+        single_design_factors = extracted_single_factors
+        # Store design factors as a list for later reuse in contrast
+        design_factors_list = copy.deepcopy(design_factors)
```

Comment:
No need for this, to remove (design_factors_list is recomputed after)

Reply:
Indeed